### PR TITLE
[REF] account,product,**: make category non require

### PR DESCRIPTION
**Pull request by mash-odoo**
** = base_import,delivery,l10n_*,stock_account,...

This commit make category field on product not required and
clean category data and demo data so user don't have to think
about what to set on category when they don't need to and
simplify category hierarchy. Also add `income_account_id and
`expense_account_id` field on company to get default income and
expense account of the company according to localization.

task-3941328

Co-Author-By: Manushi Shah <mash@odoo.com>
Co-Author-By: Dev Reshamiya <dere@odoo.com>

### CODE REVIEW

File: addons/l10n_ar/models/template_ar_base.py
Comment (kcv-odoo) [inline addons/l10n_ar/models/template_ar_base.py:15]: I think we no longer need to depend on category's accounts and change logic in `stock_account` to depend and set/get value from company's account

File: addons/mrp_account/tests/test_analytic_account.py
Comment (kcv-odoo) [inline addons/mrp_account/tests/test_analytic_account.py:36]: Why we need to set category in this test case :thinking:

File: addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py
Comment (kcv-odoo) [inline addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py:72]: same why need to set category in test case?
It is falling without it?

File: addons/point_of_sale/tests/test_point_of_sale_flow.py
Comment (kcv-odoo) [inline addons/point_of_sale/tests/test_point_of_sale_flow.py:1639]: What are you doing here :thinking:

File: addons/pos_loyalty/data/pos_loyalty_demo.xml
Comment (kcv-odoo) [inline addons/pos_loyalty/data/pos_loyalty_demo.xml:19]: unnecessary diff

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [inline addons/stock_account/models/product.py:26]: Greate then 100

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [inline addons/stock_account/models/product.py:10]:
```suggestion
VALUATION = [
```

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [inline addons/stock_account/models/product.py:15]:
```suggestion
COST_METHOD = [
```

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [inline addons/stock_account/models/product.py:144]: This is user dependent computation so maybe need to add depend_context and can be move outside loop

File: addons/stock_account/models/stock_valuation_layer.py
Comment (kcv-odoo) [inline addons/stock_account/models/stock_valuation_layer.py:67]: unnecessary diff

**Review by kcv-odoo — CHANGES_REQUESTED** [review]
Thank you for the work! :sparkles: 
Added some comments also saw there are still some test cases need category can you explain why?

File: addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py
Comment (kcv-odoo) [reply addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py:72]: :thinking:

**Review by kcv-odoo — COMMENTED** [review]
Added commits to do clean-up the approach and hopefully runbot will green :crossed_fingers: 

Added two question to look at still exploring enterprise approach :weary:

File: addons/account/models/product.py
Comment (Feyensv) [inline addons/account/models/product.py:17]: Unless python changed a lot recently, those methods do not return anything :thinking:

Comment (kcv-odoo) [reply addons/account/models/product.py:17]: oops 🤦‍♂️

File: addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py
Comment (mash-odoo) [reply addons/mrp_landed_costs/tests/test_stock_landed_costs_mrp.py:72]: When adding Landed Costs in Odoo, the product category must use FIFO or AVCO costing methods. By default, products use Standard Costing, which doesn't support landed cost adjustments and will raise an error.



To prevent this, we set the costing method for our category (categ_all) to FIFO:

`cls.product_refrigerator.categ_id.property_cost_method = 'fifo'`

File: addons/stock_landed_costs/tests/test_stock_landed_costs_rounding.py
Comment (mash-odoo) [reply addons/stock_landed_costs/tests/test_stock_landed_costs_rounding.py:23]: When adding Landed Costs, the product category must use FIFO or AVCO costing methods.

File: addons/account/models/product.py
Comment (Feyensv) [inline addons/account/models/product.py:73]:
```suggestion
                or (self.company_id or self.env.company).income_account_id
```
if the product belongs to a specific company, we should use that specific company and not the environment one.

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:27]: I understand that the valuation is configured through the group `group_stock_accounting_automatic`, but the cost_method should be configurable either on the product or on the company, we shouldn't need a category to configure those fields IMHO

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [reply addons/stock_account/models/product.py:44]: Done!

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [reply addons/stock_account/models/product.py:51]: I didn't get it. can you elaborate?

File: addons/purchase_stock/tests/test_stockvaluation.py
Comment (Feyensv) [inline addons/purchase_stock/tests/test_stockvaluation.py:307]: I dunno how this removal is related to the task here :thinking:

File: addons/purchase_stock/tests/test_stockvaluation.py
Comment (Feyensv) [inline addons/purchase_stock/tests/test_stockvaluation.py:36]: I may have missed something, but was the previous 'All' category configured in a specific way?
I feel like all the products that previously were assigned to that category should behave the same way, even without having any assigned category, so why do you assign categories to so many products in the tests who didn't have any special one before ?

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:32]:
```suggestion
    @api.depends_context('company')
    @api.depends("categ_id.property_cost_method")
```

Those two fields values won't be the same according to the company of the environment and must be correctly labeled as such (it is done automatically for company_dependent fields, but has to be done manually for standard computed fields).

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:40]:
```suggestion
    @api.depends_context('company')
    @api.depends('categ_id.property_valuation')
```

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:36]:
```suggestion
                product_template.categ_id.with_company(
                    product_template.company_id
                ).property_cost_method
```
If the product is assigned to a specific company, we should make sure the property is read for the company of the product.

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:42]:
```suggestion
        pt_with_category = self.filtered('categ_id')
```

File: addons/stock_account/models/product.py
Comment (Feyensv) [reply addons/stock_account/models/product.py:51]:
```suggestion
            product_template.valuation = product_template.categ_id.with_company(
                product_template.company_id
            ).property_valuation
```

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:236]: Same comments as for the computed fields on the templates ;)
Btw, is it really useful to redefine those fields on `product.product` model?
They could be either removed (and fall back on the template through the inherits), or be computed based on the template value instead.

**Review by Feyensv — CHANGES_REQUESTED** [review]
Hopefully last remarks/questions, it's nearly ready :pray:

Also pushed a little commit to clean up some seemingly unnecessary categories from demo data.

File: addons/purchase_stock/tests/test_stockvaluation.py
Comment (kcv-odoo) [reply addons/purchase_stock/tests/test_stockvaluation.py:307]: Not it's not related just little clean up as they are re-assigning values before using it anywhere makes this line useless :see_no_evil:

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [reply addons/stock_account/models/product.py:236]: Done!
I thought variant category can be different from template :slightly_frowning_face:

File: addons/purchase_stock/tests/test_stockvaluation.py
Comment (kcv-odoo) [reply addons/purchase_stock/tests/test_stockvaluation.py:36]: There are two reasons I had to add category on some test cases

1) Like here they update categories values of the product like change valuation, costing method in test cases that's why we need to specify category
[see line 61 in this file](https://github.com/odoo/odoo/pull/166323/files#diff-114a52640a90362788febee3f029aaf548119d5f719d99458b22f8bf7ec9766bR61)

2) In some test environment they didn't enable `stock_account.group_stock_accounting_automatic` group in test environment and if those test cases product has not category set then it'll set valuation of product to `manual_periodic` even though they change
- so either I add that group in test environment (I didn't do that because it thought adding group will impact other test case where they need `manual`) :see_no_evil: 
- or add category on product so it take valuation from category

File: addons/stock_account/models/product.py
Comment (Feyensv) [inline addons/stock_account/models/product.py:38]: No need, it is already handled in `with_company`

File: addons/purchase_stock/tests/test_stockvaluation.py
Comment (kcv-odoo) [inline addons/purchase_stock/tests/test_stockvaluation.py:1139]: This was initially 2 but it was changed to for [In PR](https://github.com/odoo/odoo/pull/110715/files) even though it should be 2 for `average`  costing method but IMO it should be 2
[it should not create extra lines for non `standard` cost method](https://github.com/odoo/odoo/blob/master/addons/purchase_stock/models/account_invoice.py#L43)

checking in detail just to be sure

**Review by Feyensv — APPROVED** [review]
LGTM (waiting for final stock approval before merge).

(and commits can be squashed)

**Comment by kcv-odoo** [comment]
Hello @odoo/rd-logistics 
Can you review this PR?

Thank you!

File: addons/stock_account/models/product.py
Comment (Whenrow) [inline addons/stock_account/models/product.py:48]: this is not related to any company record. I would add a new field `valuation` on company too to be more flexible

File: addons/stock_account/models/product.py
Comment (Whenrow) [inline addons/stock_account/models/product.py:86]: should be the one on company if the product has a company_id set on it

File: addons/stock_account/models/product.py
Comment (Whenrow) [inline addons/stock_account/models/product.py:17]: Missing conversion of following related fields from `related` to `compute` 

https://github.com/odoo/odoo/blob/60c6a52b8b9abfbad4fb47b4aaa6f22154877c80/addons/stock_account/models/stock_quant.py#L20

https://github.com/odoo/odoo/blob/60c6a52b8b9abfbad4fb47b4aaa6f22154877c80/addons/stock_account/wizard/stock_valuation_layer_revaluation.py#L39

https://github.com/odoo/odoo/blob/60c6a52b8b9abfbad4fb47b4aaa6f22154877c80/addons/stock_account/wizard/stock_valuation_layer_revaluation.py#L121

File: addons/stock_account/models/product.py
Comment (Whenrow) [inline addons/stock_account/models/product.py:139]: should be converted to computed field to get the company value if no categ

**Review by Whenrow — CHANGES_REQUESTED** [review]

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [reply addons/stock_account/models/product.py:139]: https://github.com/odoo/odoo/pull/166323#discussion_r1806814856 It will get field values from template via inherits

File: addons/stock_account/models/product.py
Comment (clesgow) [inline addons/stock_account/models/product.py:22]: Since `valuation` and `cost_method` are now non-stored computed fields, maybe we shouldn't try to set a value directly to them in the following tests anymore:
https://github.com/odoo/odoo/blob/c551b9e850bebb33a6d434cf6538e182c60d4db0/addons/purchase_stock/tests/test_stockvaluation.py#L1081-L1082
https://github.com/odoo/odoo/blob/c551b9e850bebb33a6d434cf6538e182c60d4db0/addons/stock_account/tests/test_stockvaluation.py#L3824
https://github.com/odoo/odoo/blob/c551b9e850bebb33a6d434cf6538e182c60d4db0/addons/stock_dropshipping/tests/test_purchase_order.py#L36

File: addons/stock_account/models/product.py
Comment (kcv-odoo) [reply addons/stock_account/models/product.py:48]: Hello @Whenrow 

I am not really sure about adding valuation in company to get default valuation because if group is enabled in settings means user want to use automatic valuation in their company and default will be automatic and if user want manual valuation on some product then user can change valuation on some categories to manual valuation and use that category in those products
As it was mention in specs to do that way to get default value depending on group value

Also I am not really getting good feeling about adding four field on company form view just to make category not require. Income and Expense account are okay as they are generic and used in different places but `Cost Method` and `Valuation` are specific to stock

May be we should move Cost Method in settings to get default value and check about valuation with DALA if we really want to add another option in setting or company to get default value as IMO `group_stock_accounting_automatic` Group is sufficient as valuation only contain two selection

Can you check @dalaOdoo and @Feyensv ?

Have a nice day!

Comment (Feyensv) [reply addons/stock_account/models/product.py:48]: No strong opinion on the subject, I don't mind adding a field on `res.company` to allow a more precise management of the setting depending on the company (instead of the global group).  As it's the scope of the logistics team, if they believe it's better like this, we can add it.

Comment (dalaOdoo) [reply addons/stock_account/models/product.py:48]: No need for a company setting, if no category on the product the valuation should be "manual"

Comment (kcv-odoo) [reply addons/stock_account/models/product.py:48]: @dalaOdoo And if group is enabled `group_stock_accounting_automatic` then `automatic` ?

Comment (dalaOdoo) [reply addons/stock_account/models/product.py:48]: no never automatic

**Review by Whenrow — APPROVED** [review]

**Review by Feyensv — APPROVED** [review]
LGTM :partying_face: 
Commits still have to be squashed :eyes:

@robodoo delegate+

**Comment by kcv-odoo** [comment]
Rebase the branch as there was some conflict (and waiting for upgrade team review)

**Review by ticodoo — COMMENTED** [review]
Finally had some time to look at this. I mostly have UX comments.

In addition to my comments you should update your commit message. [CLN] is not a valid commit tag as per the [guideline standards](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#tag-and-module-name) This is more of a [REF]

File: addons/product/data/product_category_demo.xml
Comment (kcv-odoo) [reply addons/product/data/product_category_demo.xml:10]: Here Clothes is in demo data and those bunch of clothes coming from data file of POS
so we can't set this category on those data
Also I am not sure creating those many categories in data file would be great idea as user will have lots of categories in their database which they want even need it and if they want to configure something for their use on category then they have to configure on each category
Like here they have add some accounts and set other things on category then we have to set same things on each category if we have lots of categories in data

File: addons/point_of_sale/data/scenarios/furniture_data.xml
Comment (kcv-odoo) [reply addons/point_of_sale/data/scenarios/furniture_data.xml]: product_category_5 was in demo data and this is data so I guess there is problem with Furniture data in POS without demo data

File: addons/account/models/company.py
Comment (kcv-odoo) [reply addons/account/models/company.py:269]: I also don't exactly know what this helper is conveying but on product category it's same I copied from there and it was there since ages so we can check with PO and update it in both places
Probably in 18.0 in another PR :thinking:

File: addons/base_import/static/csv/External_id_3rd_party_application_product_categories.csv
Comment (kcv-odoo) [reply addons/base_import/static/csv/External_id_3rd_party_application_product_categories.csv:4]: No it will not work,
As we have to set this column for all the rows or none and since this is only for documentation to display how import work and `goods`  category seems best option in data

**Comment by Feyensv** [comment]
> In addition to my comments you should update your commit message. [CLN] is not a valid commit tag as per the [guideline standards](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#tag-and-module-name) This is more of a [REF]

CLN commits are allowed by the ci/style (and should be someday in the guidelines, when we have the time to finalize their rework), but I agree this PR is more a REFactoring than a cleanup :ok_hand:

**Comment by kcv-odoo** [comment]
Hello @ticodoo :wave: 
Applied changes and added some comments. I am not really sure about adding so many categories in data I added in some places and skips some because before my changes most of them are pointing to `Salable` category so IMO `Goods` is better for now
What do you think?

**Review by ticodoo — COMMENTED** [review]
Agreed about not adding too many categories, especially since it's just for demo data. We can fix later if it's the organization isn't user friendly enough, I just figured we may as well many it as logical as possible for now with services/goods/office furniture at least being separated.

I didn't double check all the changes, but seems good to me. I thought of 1 more thing for the accounts visibility within the company form view, but it's a minor thing that can be fixed later too if a PO later confirms that it's better not to show there without any context. I'm not 100% confident about the accounting logic changes either, but if someone from the accounting team OKed it, then I guess it's fine? :shrug:

File: addons/account/views/res_company_views.xml
Comment (kcv-odoo) [reply addons/account/views/res_company_views.xml:14]: Yes we can but currently those accounts are in `Stock Valuation` section so removing them from here means they will be only accessible when stock is installed so may be we should move it some where else in accounting setting in enterprise or keep it in company from with accounting user group and remove from stock setting

For now keeping both and add accounting user group so it'll be only accessible in enterprise as before :100:

File: addons/pos_restaurant/data/scenarios/bar_data.xml
Comment (ticodoo) [inline addons/pos_restaurant/data/scenarios/bar_data.xml:6]: drinks

**Review by ticodoo — APPROVED** [review]
Good for me, just a small typo.

**Comment by mash-odoo** [comment]
@robodoo r+

**Pull request closed without merging**
